From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Thu, 1 Dec 2022 14:18:46 +0100
Subject: [PATCH] Cache world generator sea level

License: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)
Gale - https://galemc.org

* Lithium description *

Use cached sea level instead of retrieving from the registry every time.
This method is called for every block in the chunk so this will save a lot of registry lookups.
Lazy initialization added to avoid unbound registry value crash

This patch is based on the following mixin:
"net/caffeinemc/mods/lithium/mixin/gen/cached_generator_settings/NoiseBasedChunkGeneratorMixin.java"
By: SuperCoder7979 <k.pranav@gmail.com>, 2No2Name <2No2Name@web.de>
As part of: Lithium (https://github.com/CaffeineMC/lithium)
Licensed under: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)

diff --git a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
index d1068a122e5dbd5af274d0f8178ca9f5fa49f930..39705af4ec320213afce2b2c5c326a712aa5eba2 100644
--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -59,6 +59,7 @@ public final class NoiseBasedChunkGenerator extends ChunkGenerator {
     private static final BlockState AIR = Blocks.AIR.defaultBlockState();
     public final Holder<NoiseGeneratorSettings> settings;
     private final Supplier<Aquifer.FluidPicker> globalFluidPicker;
+    private int cachedSeaLevel = Integer.MIN_VALUE; // Gale - Lithium - cache world generator sea level
 
     public NoiseBasedChunkGenerator(BiomeSource biomeSource, Holder<NoiseGeneratorSettings> settings) {
         super(biomeSource);
@@ -425,7 +426,12 @@ public final class NoiseBasedChunkGenerator extends ChunkGenerator {
 
     @Override
     public int getSeaLevel() {
-        return this.settings.value().seaLevel();
+        // Gale start - Lithium - cache world generator sea level
+        if (cachedSeaLevel == Integer.MIN_VALUE) {
+            this.cachedSeaLevel = this.settings.value().seaLevel();
+        }
+        return this.cachedSeaLevel;
+        // Gale end - Lithium - cache world generator sea level
     }
 
     @Override
